File: apps/web/server/utils/initiative.ts
export const INITIATIVE_TYPES = [
  'idea',
  'hypothesis',
  'campaign',
  'opportunity',
  'decision',
  'experiment',
  'pricing',
  'market',
  'partnership',
  'internal_improvement',
] as const

export type InitiativeType = (typeof INITIATIVE_TYPES)[number]

export function requiredInitiativeType(value: unknown): InitiativeType {
  const found = typeof value === 'string' ? INITIATIVE_TYPES.find(candidate => candidate === value) : undefined
  if (!found) {
    throw createError({ statusCode: 422, statusMessage: 'Unknown initiative type' })
  }
  return found
}

export function optionalInitiativeType(value: unknown): InitiativeType | undefined {
  return value === undefined ? undefined : requiredInitiativeType(value)
}

File: apps/api/src/modules/ideas/domain/initiative.py
from typing import Final

INITIATIVE_TYPES: Final[frozenset[str]] = frozenset(
    {
        "idea",
        "hypothesis",
        "campaign",
        "opportunity",
        "decision",
        "experiment",
        "pricing",
        "market",
        "partnership",
        "internal_improvement",
    }
)


class InitiativeTypeError(ValueError):
    """Raised when an initiative type is outside the closed list."""


def validate_initiative_type(value: str) -> str:
    if value not in INITIATIVE_TYPES:
        raise InitiativeTypeError(f"Unknown initiative type: {value}")
    return value
